File: OCEL.Tests/LiteDbTests.fs
namespace OCEL.Tests

open OCEL.Types

open LiteDB
open System
open Xunit
open Xunit.Abstractions

type LiteDbTests(output: ITestOutputHelper) =

    let getCollections (db: LiteDatabase) =
        let eventsColl = db.GetCollection<string * OcelEvent>("events")
        let objectsColl = db.GetCollection<string * OcelObject>("objects")
        let globalAttrsColl = db.GetCollection<string * OcelValue>("global_attributes")
        (eventsColl, objectsColl, globalAttrsColl)

    let testCorrectNumberOfElements (db: LiteDatabase) (log: OcelLog) =
        let (e, o, a) = getCollections db
        e.Count() = log.Events.Count |> Assert.True
        o.Count() = log.Objects.Count |> Assert.True
        a.Count() = log.GlobalAttributes.Count |> Assert.True

    let log = {
            GlobalAttributes = ["version", OcelString "1.0"; "ordering", OcelString "timestamp"] |> Map.ofSeq
            Events = 
                ["e1", {
                    Activity = "Add to cart"
                    Timestamp = DateTimeOffset.Now
                    OMap = ["item1"]
                    VMap = ["price", OcelFloat 13.37; "added on", OcelTimestamp DateTimeOffset.Now] |> Map.ofList
                }; "e2", {
                    Activity = "Submit order"
                    Timestamp = DateTimeOffset.Now.AddMinutes 1
                    OMap = []
                    VMap = Map.empty
                }]
                |> Map.ofList
            Objects =
                ["item1", {
                    Type = "Item"
                    OvMap = ["Cheese", OcelInteger 12] |> Map.ofList
                }]
                |> Map.ofList
        }

    [<Fact>]
    member _.``Can serialize basic log`` () =
        let db = new LiteDatabase(":memory:")
        OCEL.OcelLiteDB.serialize db log
        testCorrectNumberOfElements db log

    [<Fact>]
    member _.``Can serialize and deserialize basic log`` () =
        let db = new LiteDatabase(":memory:")
        OCEL.OcelLiteDB.serialize db log
        testCorrectNumberOfElements db log
        let deserializedLog = OCEL.OcelLiteDB.deserialize db
        deserializedLog.IsValid |> Assert.True
        log.Equals deserializedLog |> Assert.True

    [<Fact>]
    member _.``Can serialize basic log multiple times without error`` () =
        let db = new LiteDatabase(":memory:")
        OCEL.OcelLiteDB.serialize db log
        OCEL.OcelLiteDB.serialize db log
        testCorrectNumberOfElements db log

    [<Fact>]
    member _.``Can serialize basic log to disk`` () =
        let db = new LiteDatabase(":temp:")
        OCEL.OcelLiteDB.serialize db log
        testCorrectNumberOfElements db log

    [<Fact>]
    member _.``Serialized basic log is valid`` () =
        let db = new LiteDatabase(":memory:")
        OCEL.OcelLiteDB.serialize db log
        OCEL.OcelLiteDB.validate db |> Assert.True

File: OCEL.CSharp.Tests/LiteDbTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using LiteDB;
using Xunit;
using Xunit.Abstractions;

namespace OCEL.CSharp.Tests
{
    public class LiteDbTests
    {
        private readonly ITestOutputHelper _output;

        public LiteDbTests(ITestOutputHelper output)
        {
            _output = output;
        }

        private static readonly OcelLog Log = new(
            new Dictionary<string, OcelValue>
            {
                { "version", new OcelString("1.0") },
                { "ordering", new OcelString("timestamp") }
            },
            new Dictionary<string, OcelEvent>
            {
                { "e1", new OcelEvent("Add to cart", DateTimeOffset.Now, new List<string> { "item1" }, new Dictionary<string, OcelValue>()) }
            },
            new Dictionary<string, OcelObject>
            {
                { "item1", new OcelObject("Item", new Dictionary<string, OcelValue>
                    {
                        { "Cheese", new OcelInteger(12) }

                    })
                }
            });

        private static (
            ILiteCollection<Tuple<string, OcelEvent>> e, 
            ILiteCollection<Tuple<string, OcelObject>> o, 
            ILiteCollection<Tuple<string, OcelValue>> a) 
            GetCollections(ILiteDatabase db)
        {
            var events = db.GetCollection<Tuple<string, OcelEvent>>("events");
            var objects = db.GetCollection<Tuple<string, OcelObject>>("objects");
            var attrs = db.GetCollection<Tuple<string, OcelValue>>("global_attributes");
            return (events, objects, attrs);
        }

        private static void TestCorrectNumberOfElements(ILiteDatabase db, OcelLog log)
        {
            var (e, o, a) = GetCollections(db);
            Assert.True(e.Count() == log.Events.Count);
            Assert.True(o.Count() == log.Objects.Count);
            Assert.True(a.Count() == log.GlobalAttributes.Count);
        }

        [Fact]
        public void CanDeserializeBasicLog()
        {
            var path = Path.GetFullPath(Path.Combine("..", "..", "..", "..", "..", "..", "data", "OCEL", "minimal.db"));
            var db = new LiteDatabase($"Filename={path};ReadOnly=true");
            var log = OcelLiteDB.Deserialize(db);
            Assert.True(log.IsValid);
        }

        [Fact]
        public void CanSerializeBasicLog()
        {
            var db = new LiteDatabase(":memory:");
            OcelLiteDB.Serialize(db, Log, true);
            TestCorrectNumberOfElements(db, Log);
        }

        [Fact]
        public void CanSerializeAndDeserializeBasicLog()
        {
            var db = new LiteDatabase(":memory:");
            OcelLiteDB.Serialize(db, Log, true);
            var deserializedLog = OcelLiteDB.Deserialize(db);
            TestCorrectNumberOfElements(db, Log);
            Assert.True(deserializedLog.IsValid);
        }

        [Fact]
        public void CanDeserializeBasicLogMultipleTimesWithoutError()
        {
            var db = new LiteDatabase(":memory:");
            OcelLiteDB.Serialize(db, Log, true);
            OcelLiteDB.Serialize(db, Log, true);
            TestCorrectNumberOfElements(db, Log);
        }

        [Fact]
        public void CanSerializeBasicLogToDisk()
        {
            var db = new LiteDatabase(":temp:");
            OcelLiteDB.Serialize(db, Log, true);
            TestCorrectNumberOfElements(db, Log);
        }

        [Fact]
        public void SerializedBasicLogIsValid()
        {
            var db = new LiteDatabase(":memory:");
            OcelLiteDB.Serialize(db, Log, true);
            Assert.True(OcelLiteDB.Validate(db));
        }
    }
}
